File: src/main/java/com/rustem/simplecrudapplication/service/CustomerServiceImpl.java
package com.rustem.simplecrudapplication.service;

import com.rustem.simplecrudapplication.model.Customer;
import com.rustem.simplecrudapplication.repository.CustomerRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Slf4j
@Service
@Transactional
public class CustomerServiceImpl implements CustomerService {

    private final CustomerRepository customerRepository;

    @Autowired
    public CustomerServiceImpl(CustomerRepository customerRepository) {
        this.customerRepository = customerRepository;
    }

    @Transactional(readOnly = true)
    @Override
    public Customer getById(Long id) {
        log.info("In CustomerServiceImpl getById {}", id);
        Optional<Customer> customer = customerRepository.findById(id);
        return customer.orElse(null);
    }

    @Override
    public void save(Customer customer) {
        log.info("In CustomerServiceImpl save {}", customer);
        customerRepository.save(customer);
    }

    @Override
    public void update(Customer c) {

        Optional<Customer> customer = customerRepository.findById(c.getId());

        Integer update = null;
        if (customer.isPresent()) {
            update = customerRepository.update(c.getId(), c.getFirstName(), c.getLastName(), c.getAddress(), c.getBudget());
            log.info("Updated: {}", update);
        }
    }

    @Override
    public void delete(Long id) {
        log.info("In CustomerServiceImpl delete {}", id);
        customerRepository.deleteById(id);
    }

    @Override
    public List<Customer> getAll() {
        log.info("In CustomerServiceImpl getAll");
        return customerRepository.findAll();
    }
}

File: src/main/java/com/rustem/simplecrudapplication/service/CustomerServiceImpl.kt
package com.rustem.simplecrudapplication.service

import com.rustem.simplecrudapplication.model.Customer
import com.rustem.simplecrudapplication.repository.CustomerRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class CustomerServiceImpl(private val customerRepository: CustomerRepository) : CustomerService {

    @Transactional(readOnly = true)
    override fun getById(id: Long): Customer? {
        log.info("In CustomerServiceImpl getById {}", id)
        val customer: Optional<Customer?> = customerRepository.findById(id)
        return customer.orElse(null)
    }

    override fun save(customer: Customer) {
        log.info("In CustomerServiceImpl save {}", customer)
        customerRepository.save<Customer>(customer)
    }

    override fun update(c: Customer) {
        val customer: Optional<Customer?> = customerRepository.findById(c.id!!)
        var update: Int? = null
        if (customer.isPresent) {
            update =
                customerRepository.update(c.id, c.firstName, c.lastName, c.address, c.budget)
            log.info("Updated: {}", update)
        }
    }

    override fun delete(id: Long) {
        log.info("In CustomerServiceImpl delete {}", id)
        customerRepository.deleteById(id)
    }

    override fun getAll(): MutableList<Customer?> {
        log.info("In CustomerServiceImpl getAll")
        return customerRepository.findAll()
    }

    companion object {
        private val log = LoggerFactory.getLogger(CustomerServiceImpl::class.java)
    }
}